File: python_examples/index_operations.py
#!/usr/bin/env python3
"""
Index Operations Example

Demonstrates key features of the enhanced MF4 index system in Python.
"""

import mf4_rs
import os

def main():
    print("🚀 Index Operations Example")
    print("=" * 40)
    
    # Create a simple test file
    mdf_file = "index_example.mf4"
    index_file = "index_example.json"
    
    try:
        # Clean up existing files
        for f in [mdf_file, index_file]:
            if os.path.exists(f):
                os.remove(f)
        
        print("1️⃣ Creating test MDF file...")
        create_simple_mdf(mdf_file)
        
        print("2️⃣ Creating enhanced index...")
        # Create enhanced index - this resolves all conversions automatically
        index = mf4_rs.PyMdfIndex.from_file(mdf_file)
        index.save_to_file(index_file)
        
        # Check if it's enhanced
        has_resolved = index.has_resolved_conversions()
        print(f"   Enhanced conversions: {has_resolved}")
        
        print("3️⃣ Using the enhanced index...")
        # Load from JSON
        loaded_index = mf4_rs.PyMdfIndex.load_from_file(index_file)
        
        # List available data
        groups = loaded_index.list_channel_groups()
        print(f"   Channel groups: {len(groups)}")
        
        for group_idx, group_name, channel_count in groups:
            print(f"   Group {group_idx}: {channel_count} channels")
            
            channels = loaded_index.list_channels(group_idx)
            if channels:
                for ch_idx, ch_name, data_type in channels:
                    print(f"     - {ch_name} ({data_type.name})")
        
        print("4️⃣ Reading data efficiently...")
        # Read by name - fastest way
        temp_values = loaded_index.read_channel_values_by_name("Temperature", mdf_file)
        print(f"   Temperature: {len(temp_values)} values")
        print(f"   Range: {temp_values[0]} to {temp_values[-1]}")
        
        print("5️⃣ HTTP optimization features...")
        # Find channel location
        temp_location = loaded_index.find_channel_by_name("Temperature")
        if temp_location:
            group_idx, channel_idx = temp_location
            
            # Get byte range info
            total_bytes, range_count = loaded_index.get_channel_byte_summary(group_idx, channel_idx)
            print(f"   Temperature data: {total_bytes} bytes in {range_count} ranges")
            
            # Get byte ranges for partial reading
            partial_ranges = loaded_index.get_channel_byte_ranges_for_records(group_idx, channel_idx, 0, 5)
            partial_bytes = sum(length for _, length in partial_ranges)
            savings = (1 - partial_bytes / total_bytes) * 100
            
            print(f"   First 5 records: {partial_bytes} bytes ({savings:.1f}% bandwidth savings)")
            print(f"   HTTP range: bytes={partial_ranges[0][0]}-{partial_ranges[0][0] + partial_ranges[0][1] - 1}")
        
        print("\n✅ Enhanced index features demonstrated!")
        print("\n🎯 Key Benefits:")
        print("   • Index contains all data needed for conversions")
        print("   • Perfect for HTTP/remote file access")
        print("   • Precise byte range calculations")
        print("   • Fast name-based lookups")
        print("   • Much smaller than original MDF files")
        
        # Show file sizes
        mdf_size = os.path.getsize(mdf_file)
        index_size = os.path.getsize(index_file)
        print(f"\n📊 File Sizes:")
        print(f"   MDF:   {mdf_size:,} bytes")
        print(f"   Index: {index_size:,} bytes ({index_size/mdf_size*100:.1f}% of original)")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
    
    finally:
        # Cleanup
        for f in [mdf_file, index_file]:
            if os.path.exists(f):
                os.remove(f)

def create_simple_mdf(file_path):
    """Create a simple MDF file for testing."""
    writer = mf4_rs.PyMdfWriter(file_path)
    writer.init_mdf_file()
    
    # Create channel group
    group = writer.add_channel_group("Test Data")
    
    # Add channels
    time_ch = writer.add_time_channel(group, "Time")
    temp_ch = writer.add_float_channel(group, "Temperature")
    rpm_ch = writer.add_int_channel(group, "RPM")
    
    # Write data
    writer.start_data_block(group)
    
    for i in range(20):
        time_val = mf4_rs.create_float_value(i * 0.1)
        temp_val = mf4_rs.create_float_value(20 + i * 2.5)  # Temperature 20-67.5°C
        rpm_val = mf4_rs.create_uint_value(1000 + i * 50)  # RPM 1000-1950
        
        writer.write_record(group, [time_val, temp_val, rpm_val])
    
    writer.finish_data_block(group)
    writer.finalize()

if __name__ == "__main__":
    main()

File: examples/index_operations.rs
use mf4_rs::blocks::common::DataType;
use mf4_rs::error::MdfError;
use mf4_rs::index::MdfIndex;
use mf4_rs::parsing::decoder::DecodedValue;
use mf4_rs::writer::MdfWriter;

fn main() -> Result<(), MdfError> {
    let mdf_file = "index_example.mf4";
    let index_file = "index_example.json";

    println!("=== Index Operations Example ===");

    // Step 1: Create an MDF file with some test data
    println!("📄 Creating MDF file with test data...");
    create_test_mdf_file(mdf_file)?;

    // Step 2: Create enhanced index that resolves all conversion dependencies
    println!("📇 Creating enhanced index with resolved conversions...");
    let index = MdfIndex::from_file(mdf_file)?;
    index.save_to_file(index_file)?;

    println!("   ✅ Index created and saved to '{}'", index_file);

    // Step 3: Load the index and demonstrate self-contained conversion capability
    println!("🔄 Loading index and testing self-contained conversions...");
    let loaded_index = MdfIndex::load_from_file(index_file)?;

    // Step 4: Read and convert data using only the index + a bound data source.
    println!("📊 Reading channel values via enhanced index...");

    // List available channels by navigating the index by name
    println!("\nAvailable channels:");
    for channel in &loaded_index.groups()[0].channels {
        println!(
            "  {} ({:?})",
            channel.name.as_deref().unwrap_or("<unnamed>"),
            channel.data_type
        );
    }

    // Bind the data source once, then read by channel name.
    let mut data = loaded_index.open_file(mdf_file)?;
    let values = data.values("Time")?;

    println!("\nChannel values read from enhanced index:");
    for (i, value) in values.iter().enumerate().take(10) {
        println!("  Record {}: {:?}", i, value);
    }
    if values.len() > 10 {
        println!("  ... and {} more records", values.len() - 10);
    }

    // Step 5: Demonstrate byte range efficiency
    println!("\n🎯 Byte Range Analysis (Partial Reading):");
    let byte_ranges = loaded_index.byte_ranges("Time")?;
    let total_bytes: u64 = byte_ranges.iter().map(|(_, len)| len).sum();

    println!("  Total bytes needed: {}", total_bytes);
    println!("  Number of byte ranges: {}", byte_ranges.len());
    println!("  Ranges: {:?}", byte_ranges);

    // Step 6: Read another channel by name
    println!("\n🏷️  Reading 'Temperature' by name:");
    let temp_values = data.values("Temperature")?;
    println!("  Read {} values for channel 'Temperature'", temp_values.len());

    // Step 7: Show resolved conversion information
    println!("\n🔧 Conversion Resolution Status:");
    let group = &loaded_index.groups()[0];
    for (i, channel) in group.channels.iter().enumerate() {
        println!(
            "  Channel {}: {}",
            i,
            channel.name.as_deref().unwrap_or("<unnamed>")
        );
        if let Some(conversion) = &channel.conversion {
            println!("    Conversion type: {:?}", conversion.cc_type);
            if let Some(resolved_texts) = &conversion.resolved_texts {
                println!("    Resolved texts: {} entries", resolved_texts.len());
            }
            if let Some(resolved_conversions) = &conversion.resolved_conversions {
                println!(
                    "    Resolved nested conversions: {} entries",
                    resolved_conversions.len()
                );
            }
        } else {
            println!("    No conversion block");
        }
    }

    println!("\n🎉 Enhanced index example completed successfully!");
    println!("\nKey Benefits Demonstrated:");
    println!("  ✅ All conversion dependencies resolved during index creation");
    println!("  ✅ No file access needed for conversion operations during data reading");
    println!("  ✅ Perfect for HTTP/remote file access scenarios");
    println!("  ✅ Maintains full compatibility with existing functionality");
    println!("  ✅ Works with all conversion types: linear, text, rational, algebraic, etc.");

    // Clean up
    std::fs::remove_file(mdf_file).ok();
    std::fs::remove_file(index_file).ok();

    Ok(())
}

fn create_test_mdf_file(path: &str) -> Result<(), MdfError> {
    let mut writer = MdfWriter::new(path)?;
    writer.init_mdf_file()?;

    let cg_id = writer.add_channel_group(None, |_| {})?;

    // Create a time channel (master)
    let time_ch_id = writer.add_channel(&cg_id, None, |ch| {
        ch.data_type = DataType::FloatLE;
        ch.name = Some("Time".to_string());
        ch.bit_count = 64;
    })?;
    writer.set_time_channel(&time_ch_id)?;

    // Create a temperature channel with linear conversion
    writer.add_channel(&cg_id, Some(&time_ch_id), |ch| {
        ch.data_type = DataType::UnsignedIntegerLE;
        ch.name = Some("Temperature".to_string());
        ch.bit_count = 16;
    })?;

    // Create a status channel (could be enhanced with text conversion later)
    writer.add_channel(&cg_id, Some(&time_ch_id), |ch| {
        ch.data_type = DataType::UnsignedIntegerLE;
        ch.name = Some("Status".to_string());
        ch.bit_count = 8;
    })?;

    // Write sample data
    writer.start_data_block_for_cg(&cg_id, 0)?;

    for i in 0..20 {
        let time = i as f64 * 0.1;
        let temp_raw = (i * 10 + 200) as u64; // Raw ADC value
        let status = if i % 5 == 0 { 1u64 } else { 0u64 };

        writer.write_record(
            &cg_id,
            &[
                DecodedValue::Float(time),
                DecodedValue::UnsignedInteger(temp_raw),
                DecodedValue::UnsignedInteger(status),
            ],
        )?;
    }

    writer.finish_data_block(&cg_id)?;
    writer.finalize()?;

    Ok(())
}
